test/unit/math/opencl/kernel_generator/select_test.cpp

```diff
@@ -0,0 +1,126 @@
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/copy.hpp>
+#include <test/unit/math/opencl/kernel_generator/reference_kernel.hpp>
+#include <Eigen/Dense>
+#include <gtest/gtest.h>
+#include <string>
+
+using Eigen::Matrix;
+using Eigen::MatrixXd;
+using Eigen::MatrixXi;
+using stan::math::matrix_cl;
+
+#define EXPECT_MATRIX_NEAR(A, B, DELTA) \
+  for (int i = 0; i < A.size(); i++)    \
+    EXPECT_NEAR(A(i), B(i), DELTA);
+
+TEST(MathMatrixCL, select_errors) {
+  using stan::math::select;
+
+  matrix_cl<double> m_double(2, 3);
+  matrix_cl<double> m_rows(3, 3);
+  matrix_cl<double> m_cols(2, 4);
+  matrix_cl<int> m_int(2, 3);
+  matrix_cl<bool> m_bool(2, 3);
+
+  EXPECT_NO_THROW(select(m_bool, m_double, m_double));
+  EXPECT_NO_THROW(select(m_bool, m_double, m_int));
+  EXPECT_NO_THROW(select(m_bool, m_int, m_double));
+  EXPECT_NO_THROW(select(m_int, m_double, m_double));
+  EXPECT_THROW(select(m_rows, m_double, m_double), std::invalid_argument);
+  EXPECT_THROW(select(m_cols, m_double, m_double), std::invalid_argument);
+  EXPECT_THROW(select(m_bool, m_rows, m_double), std::invalid_argument);
+  EXPECT_THROW(select(m_bool, m_cols, m_double), std::invalid_argument);
+  EXPECT_THROW(select(m_bool, m_double, m_rows), std::invalid_argument);
+  EXPECT_THROW(select(m_bool, m_double, m_cols), std::invalid_argument);
+}
+
+TEST(MathMatrixCL, select_test) {
+  std::string kernel_filename = "select.cl";
+  using stan::math::select;
+  MatrixXd m1(2, 3);
+  m1 << 1, 2.5, 3, 4, 5, 6.3;
+  MatrixXd m2(2, 3);
+  m2 << 10, 100, 1000, 30.5, 12.1, 0.3;
+  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond(2, 3);
+  m_cond << true, false, false, true, true, false;
+
+  matrix_cl<double> m1_cl(m1);
+  matrix_cl<double> m2_cl(m2);
+  matrix_cl<bool> m_cond_cl(m_cond);
+
+  auto tmp = select(m_cond_cl, m1_cl, m2_cl);
+
+  matrix_cl<double> res_cl;
+  std::string kernel_src = tmp.get_kernel_source_for_evaluating_into(res_cl);
+  stan::test::store_reference_kernel_if_needed(kernel_filename, kernel_src);
+  std::string expected_kernel_src
+      = stan::test::load_reference_kernel(kernel_filename);
+  EXPECT_EQ(expected_kernel_src, kernel_src);
+
+  res_cl = tmp;
+  MatrixXd res = stan::math::from_matrix_cl(res_cl);
+
+  MatrixXd correct = m_cond.select(m1, m2);
+  EXPECT_MATRIX_NEAR(res, correct, 1e-9);
+}
+
+TEST(MathMatrixCL, multiple_operations) {
```

Comment:
Mind also adding a zero-size test here? Something like: 
```cpp
TEST(MathMatrixCL, multiple_operations) {
  using stan::math::select;
  MatrixXd m0(0, 0);
  MatrixXd m1(2, 3);
  m1 << 1, 2.5, 3, 4, 5, 6.3;
  MatrixXi m2(2, 3);
  m2 << 10, 100, 1000, 30, 12, 0;
  MatrixXd m3(2, 3);
  m3 << 1.4, 12.5, 6.3, 34.4, -5, -6.3;
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond0(0, 0);
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond(2, 3);
  m_cond << true, false, false, true, true, false;
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond2(2, 3);
  m_cond2 << false, false, true, true, false, true;

  matrix_cl<double> m0_cl(m0);
  matrix_cl<double> m1_cl(m1);
  matrix_cl<int> m2_cl(m2);
  matrix_cl<double> m3_cl(m3);
  matrix_cl<bool> m_cond_cl(m_cond);
  matrix_cl<bool> m_cond2_cl(m_cond2);
  matrix_cl<bool> m_cond0_cl(m_cond0);
  auto tmp = select(m_cond_cl, m1_cl, select(m_cond2_cl, m2_cl, m3_cl));
  matrix_cl<double> res_cl = tmp;

  MatrixXd res = stan::math::from_matrix_cl(res_cl);

  MatrixXd correct = m_cond.select(m1, m_cond2.select(m2.cast<double>(), m3));
  EXPECT_MATRIX_NEAR(res, correct, 1e-9);
  
  auto tmp0 = select(m_cond0_cl, m0_cl, select(m_cond0_cl, m0_cl, m0_cl));
  matrix_cl<double> res0_cl = tmp0;

  MatrixXd res0 = stan::math::from_matrix_cl(res0_cl);

  MatrixXd correct0 = m_cond0.select(m0, m_cond0.select(m0.cast<double>(), m0));
  EXPECT_MATRIX_NEAR(res0, correct0, 1e-9);
}
```

Reply:
I simplified your suggestion and put it into a separate test.

Reply:
Great.